providers/implementations/signature/sm2sig.c

```diff
@@ -0,0 +1,501 @@
+#include <openssl/params.h>
+#include <openssl/evp.h>
+#include <openssl/err.h>
+#include "internal/nelem.h"
+#include "internal/sizes.h"
+#include "prov/providercommonerr.h"
+#include "prov/implementations.h"
+#include "prov/providercommonerr.h"
+#include "prov/provider_ctx.h"
+#include "crypto/ec.h"
+#include "crypto/sm2.h"
+
+static OSSL_OP_signature_newctx_fn sm2sig_newctx;
+static OSSL_OP_signature_sign_init_fn sm2sig_signature_init;
+static OSSL_OP_signature_verify_init_fn sm2sig_signature_init;
+static OSSL_OP_signature_sign_fn sm2sig_sign;
+static OSSL_OP_signature_verify_fn sm2sig_verify;
+static OSSL_OP_signature_digest_sign_init_fn sm2sig_digest_signverify_init;
+static OSSL_OP_signature_digest_sign_update_fn sm2sig_digest_signverify_update;
+static OSSL_OP_signature_digest_sign_final_fn sm2sig_digest_sign_final;
+static OSSL_OP_signature_digest_verify_init_fn sm2sig_digest_signverify_init;
+static OSSL_OP_signature_digest_verify_update_fn sm2sig_digest_signverify_update;
+static OSSL_OP_signature_digest_verify_final_fn sm2sig_digest_verify_final;
+static OSSL_OP_signature_freectx_fn sm2sig_freectx;
+static OSSL_OP_signature_dupctx_fn sm2sig_dupctx;
+static OSSL_OP_signature_get_ctx_params_fn sm2sig_get_ctx_params;
+static OSSL_OP_signature_gettable_ctx_params_fn sm2sig_gettable_ctx_params;
+static OSSL_OP_signature_set_ctx_params_fn sm2sig_set_ctx_params;
+static OSSL_OP_signature_settable_ctx_params_fn sm2sig_settable_ctx_params;
+static OSSL_OP_signature_get_ctx_md_params_fn sm2sig_get_ctx_md_params;
+static OSSL_OP_signature_gettable_ctx_md_params_fn sm2sig_gettable_ctx_md_params;
+static OSSL_OP_signature_set_ctx_md_params_fn sm2sig_set_ctx_md_params;
+static OSSL_OP_signature_settable_ctx_md_params_fn sm2sig_settable_ctx_md_params;
+
+/*
+ * What's passed as an actual key is defined by the KEYMGMT interface.
+ * We happen to know that our KEYMGMT simply passes EC structures, so
+ * we use that here too.
+ */
+typedef struct {
+    OPENSSL_CTX *libctx;
+    EC_KEY *ec;
+
+    /*
+     * Flag to determine if the hash function can be changed (1) or not (0)
+     * Because it's dangerous to change during a DigestSign or DigestVerify
+     * operation, this flag is cleared by their Init function, and set again
+     * by their Final function.
+     */
+    unsigned int flag_allow_md : 1;
+
+    char mdname[OSSL_MAX_NAME_SIZE];
+
+    /* The Algorithm Identifier of the combined signature agorithm */
+    unsigned char aid[OSSL_MAX_ALGORITHM_ID_SIZE];
+    size_t  aid_len;
+
+    /* main digest */
+    EVP_MD *md;
+    EVP_MD_CTX *mdctx;
+    size_t mdsize;
+} PROV_SM2_CTX;
+
+/* TODO: it seems SM2 doesn't need this */
+static int sm2sig_get_md_nid(const EVP_MD *md)
+{
+    /*
+     * Because the EC library deals with NIDs, we need to translate.
+     * We do so using EVP_MD_is_a(), and therefore need a name to NID
+     * map.
+     */
+    static const OSSL_ITEM name_to_nid[] = {
+        { NID_sm2,   OSSL_DIGEST_NAME_SM3 },
```

Comment:
.. this wasn't quite right

Reply:
This still isn't quite right

Reply:
Hmmm? I thought I have fixed this...Did you comment on an outdated code?